src/app/dashboard/dashboard.controller.js

```diff
@@ -18,61 +19,100 @@
+        locals: { plugin: copy, isUpdate: isUpdate }
       });
     }
 
     function PluginConfigCtrl() {
       var vm = this;
       var pluginTitle = vm.plugin.title;
+      var isUpdate = vm.isUpdate;
+
+      if (isUpdate) {
+        vm.save = save;
+      } else {
+        // Don't allow `save` until we have a unique name to suggest.
+        getUniquePlugin(vm).then(function() {
+          vm.save = save;
+        });
+
+      }
 
       vm.close = $mdDialog.hide;
-      vm.save = save;
 
       function save() {
-        return $dashboard
-          .updatePlugin(pluginTitle, vm.plugin)
-          .then(vm.close);
+        var id = vm.id || getSlug(pluginTitle);
+        var $plugin = $dashboard.getPlugin(id);
+
+        if (isUpdate) {
+          // We don't have to find a unique plugin ID.
+          return $plugin.save(vm.plugin).then(vm.close);
+        } else {
+          return getUniquePlugin(vm)
+            .then(function($plugin) {
+              return $plugin.save(vm.plugin);
+            })
+            .then(vm.close);
+        }
       }
     }
 
-    function clone(plugin) {
-      return configure(plugin, false);
-    }
+    function getUniquePlugin(vm) {
+      var id = vm.id || getSlug(vm.plugin.title);
+      var originalTitle = vm.plugin.title;
 
-    function getUniqueName(name) {
-      if (!$dashboard.getPluginByTitle(name)) {
-        return name;
-      }
+      return $dashboard.read().then(function(data) {
+        function exists(id) {
+          return data && data.plugins && data.plugins[id];
+        }
 
-      var i = 0;
-      var title;
+        if (!exists(id)) {
+          // This name is fine.
+          return $dashboard.getPlugin(id);
+        } else {
+          var i = 0;
 
-      while (++i) {
-        title = name + ' ' + i;
+          while (++i) {
+            vm.plugin.title = originalTitle + ' (' + i + ')';
+            id = getSlug(vm.plugin.title);
 
-        if (!$dashboard.getPluginByTitle(title)) {
-          return title;
+            if (!exists(id)) {
+              return $dashboard.getPlugin(id);
+            }
+          }
         }
-      }
+      });
     }
 
-    function remove(plugin) {
-      return $dashboard
-        .removePlugin(plugin)
+    function clone(pluginId, pluginData) {
+      return configure(pluginData, false);
+    }
+
+    function remove(pluginId, pluginData) {
+      var $plugin = $dashboard.getPlugin(pluginId);
+
+      return $plugin.delete()
+        .then(function() {
+          // Return to the dashboard view.
+          return $state.go('dashboard', {
+            dashboardId: $dashboard.id
+          });
+        })
         .then(showUndoToast)
         .then(function(action) {
           if (action === 'ok') {
-            return $dashboard.addPlugin(plugin);
+            // Go back to the plugin's page.
+            return $plugin.save(pluginData)
+              .then(function() {
+                return $state.go('plugin', {
```

Comment:
I kept this in. If we think it's weird after using it for a while, it's easy to pull it out.